crates/store/re_query2/src/latest_at.rs

```diff
@@ -0,0 +1,681 @@
+                    // recursive flag is set.
+                    #[allow(clippy::collapsible_if)] // readability
+                    if clear_entity_path == *entity_path || found_recursive_clear {
+                        if let Some(index) = cached.index(&query.timeline()) {
+                            if compare_indices(index, max_clear_index)
+                                == std::cmp::Ordering::Greater
+                            {
+                                max_clear_index = index;
+                            }
+                        }
+                    }
+                }
+
+                let Some(parent_entity_path) = clear_entity_path.parent() else {
+                    break;
+                };
+
+                clear_entity_path = parent_entity_path;
+            }
+        }
+
+        for component_name in component_names {
+            let key = CacheKey::new(entity_path.clone(), query.timeline(), component_name);
+
+            let cache = Arc::clone(
+                self.latest_at_per_cache_key
+                    .write()
+                    .entry(key.clone())
+                    .or_insert_with(|| Arc::new(RwLock::new(LatestAtCache::new(key.clone())))),
+            );
+
+            let mut cache = cache.write();
+            cache.handle_pending_invalidation();
+            if let Some(cached) = cache.latest_at(store, query, entity_path, component_name) {
+                // 1. A `Clear` component doesn't shadow its own self.
+                // 2. If a `Clear` component was found with an index greater than or equal to the
+                //    component data, then we know for sure that it should shadow it.
+                if let Some(index) = cached.index(&query.timeline()) {
+                    if component_name == ClearIsRecursive::name()
+                        || compare_indices(index, max_clear_index) == std::cmp::Ordering::Greater
+                    {
+                        results.add(component_name, index, cached);
+                    }
+                }
+            }
+        }
+
+        results
+    }
+}
+
+// --- Results ---
+
+/// Results for a latest-at query.
+///
+/// Use [`LatestAtResults::get`] and/or [`LatestAtResults::get_required`] in order to access
+/// the results for each individual component.
+#[derive(Debug)]
+pub struct LatestAtResults {
+    /// The associated [`EntityPath`].
+    pub entity_path: EntityPath,
+
+    /// The query that yielded these results.
+    pub query: LatestAtQuery,
+
+    /// The compound index of this query result.
+    ///
+    /// A latest-at query is a compound operation that gathers data from many different rows.
+    /// The index of that compound result corresponds to the index of most the recent row in all the
+    /// sub-results, as defined by time and row-id order.
+    pub compound_index: (TimeInt, RowId),
+
+    /// Results for each individual component.
+    pub components: IntMap<ComponentName, UnitChunkShared>,
```

Comment:
Materializing chunks for the results was a bit unexpected for me here. I still would have expected this to be something like a `UnitChunkSlice`, though in practice I'm not sure the distinction matters as much as it does in my head.

Reply:
Also probably worth clarifying if UnitChunkShared MUST always contain the corresponding component, or if we are allowed to end up in situations where the component is included in the IntMap, but UnitChunkShared doesn't know about it.

Reply:
> Materializing chunks for the results was a bit unexpected for me here. I still would have expected this to be something like a `UnitChunkSlice`, though in practice I'm not sure the distinction matters as much as it does in my head.

I'm not sure what you mean by `UnitChunkSlice`?

Do you mean that the different Chunks should be just be slices into Chunks in the store? If so, that's already the case.

If you instead mean that there shouldn't be a map at all, but a rather a single Chunk that represents the merged data of all the Chunks -- I don't think so. It's quite powerful to be able to access the original chunks of each component,

The caller of the API won't have to worry about this in most cases anyhow, the higher-level APIs on `LatestAtResults` transparently take care of this.

Reply:
> Do you mean that the different Chunks should be just be slices into Chunks in the store? If so, that's already the case.

I think my point is that I expected this to be a bit more explicit. A chunk slice would be something like `ChunkSlice = (Chunk, IndexMin, IndexMax)`